File: typescript/src/useCase/OrderApprovalUseCase.ts
import Order from '../domain/Order';
import {OrderStatus} from '../domain/OrderStatus';
import OrderRepository from '../repository/OrderRepository';
import OrderApprovalRequest from './OrderApprovalRequest';

class OrderApprovalUseCase {
  private readonly orderRepository: OrderRepository;

  public constructor(orderRepository: OrderRepository){
    this.orderRepository = orderRepository;
  }

  public run(request: OrderApprovalRequest): void {
    const order: Order = this.orderRepository.getById(request.getOrderId());

    order.verifyApprovalStatus(request);

    order.setStatus(request.isApproved() ? OrderStatus.APPROVED : OrderStatus.REJECTED);
    this.orderRepository.save(order);
  }
}

export default OrderApprovalUseCase;

File: python/src/use_case/order_approval_use_case.py
from .order_approval_request import OrderApprovalRequest
from ..repository.order_repository import OrderRepository


class OrderApprovalUseCase:
    def __init__(self, order_repository: OrderRepository) -> None:
        self._order_repository = order_repository

    def run(self, request: OrderApprovalRequest) -> None:
        order = self._order_repository.get_by_id(request.order_id)

        order.approve(request.approved)

        self._order_repository.save(order)
